File: ABC/284/C.go
package main

import "fmt"

func main() {
	var n, m int
	fmt.Scan(&n, &m)
	g := make([][]int, n)

	for i := 0; i < m; i++ {
		var u, v int
		fmt.Scan(&u, &v)
		u--
		v--
		g[u] = append(g[u], v)
		g[v] = append(g[v], u)
	}

	visited := make([]bool, n)
	var cnt int
	for i := 0; i < n; i++ {
		if visited[i] {
			continue
		}
		cnt++
		dfs(&g, &visited, i)
	}
	fmt.Println(cnt)
}

func dfs(g *[][]int, visited *[]bool, i int) {
	(*visited)[i] = true
	for j := 0; j < len((*g)[i]); j++ {
		if (*visited)[(*g)[i][j]] {
			continue
		}
		dfs(g, visited, (*g)[i][j])
	}
}

File: ABC/284/c.rs
use proconio::{input, marker::Usize1};

fn main() {
    input! {
        n: usize,
        m: usize,
        uv: [(Usize1, Usize1); m],
    }
    let mut g = vec![Vec::new(); n];
    for (u, v) in uv {
        g[u].push(v);
        g[v].push(u);
    }

    let mut cnt  = 0;
    let mut visited = vec![false; n];

    for i in 0..n {
        if dfs(i, &g, &mut visited) {
            cnt += 1;
        }
    }
    println!("{}", cnt);
}

fn dfs(i: usize, g: &[Vec<usize>], visited: &mut [bool]) -> bool {
    if visited[i] {
        return false;
    }
    visited[i] = true;
    for j in &g[i] {
        dfs(*j, g, visited);
    }
    true
}
